components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarAutocompleteFeature.kt

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.toolbar
+
+import mozilla.components.browser.domains.DomainAutocompleteProvider
+import mozilla.components.concept.storage.HistoryStorage
+import mozilla.components.concept.toolbar.AutocompleteResult
+import mozilla.components.concept.toolbar.Toolbar
+
+/**
+ * Feature implementation for connecting a toolbar with a list of autocomplete providers.
+ */
+class ToolbarAutocompleteFeature(val toolbar: Toolbar) {
+    private val historyProviders: MutableList<HistoryStorage> = mutableListOf()
+    private val domainProviders: MutableList<DomainAutocompleteProvider> = mutableListOf()
+    private val allProviders = { historyProviders.asSequence().plus(domainProviders) }
+
+    init {
+        toolbar.setAutocompleteListener { value, acceptor ->
+            acceptor?.let {
+                for (provider in allProviders()) {
+                    val result = provider.autocomplete(value)
+                    if (result != null) {
+                        it.applyAutocompleteResult(result)
+                        return@setAutocompleteListener
+                    }
+                }
+                it.noAutocompleteResult()
+            }
+        }
+    }
+
+    fun addHistoryStorageProvider(provider: HistoryStorage) {
+        historyProviders.add(provider)
+    }
+
+    fun addDomainProvider(provider: DomainAutocompleteProvider) {
+        domainProviders.add(provider)
+    }
+
+    // The rest is a little bit of boilerplate to tie things together.
+    // We terminate our dependency chains at the 'feature' level. This is an alternative to creating
+    // an "AutocompleteProvider" concept. This pattern allows a storage component, for example, to
+    // be standalone and not depend on something like 'concept-domains'.
+    // This pattern also helps prevent circular dependency situations from arising over time.
+
```

Comment:
I get the reason we're doing this, but it is a bit unfortunate :). Wondering if we should create a single AutocompleteResult (instead of 3-4 distinct types) in a root module e.g. support-base? I think this would simplify this quite a bit!

Reply:
@csadilek One of the reasons *for* doing this boilerplate that's not currently obvious is that it enables individual providers to define a "result" as they please, and expose a greater set of functionality as they see fit. The `feature` component is then responsible for massaging different results together. If we have a common definition in `support-*`, we introduce an implicit dependency between all of our providers.

The other side of having this implicit dependency is that it has a benefit of imposing a "growing in lockstep" limitation on the different provider implementations, which could be quite useful.

Reply:
Boilerplate is the price for abstractions and a layered architecture. This seems fine to me here since Features are translating between two worlds. Anyhow, boilerplate is something we can always optimize later. :)

Reply:
Yeah, if the results were different I wouldn't have noticed. It's just that it's the same type with 4 copies, which tipped me off. So, this boilerplate is caused by a lack of abstraction (over the result) not because of it. :)

Anyways, like I wrote, I don't feel strongly. Just a comment...